### PR TITLE
Component Grouping

### PR DESCRIPTION
**Pull request by Aceeri**
fixes https://github.com/slide-rs/specs/issues/155

Restrutures into a workspace of `specs` and `specs_derive`.

### CODE REVIEW

File: specs/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# specs
```
Comment (torkleyy) [inline specs/README.md]: It seems your rebase wasn't quite correct, this is the old README.

File: specs_derive/src/component_group.rs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    // Normal group methods.
+    let default = quote! {
+        fn local_components() -> Vec<&'static str> {
+            vec![ #( stringify!(#comp_ty), )* ]
+        }
+        fn components() -> Vec<&'static str> {
+            #[allow(unused_mut)]
+            let mut list = #name::local_components();
+            #(
+                for component in #subgroup_ty::components() {
```
Comment (WaDelma) [inline specs_derive/src/component_group.rs]: Can't this use `extend` instead of manual loop?

File: specs/src/group.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    /// Registers the components into the world.
+    fn register(&mut World);
+}
+
+/// Group of serializable components.
+#[cfg(feature="serialize")]
+pub trait SerializeGroup: ComponentGroup {
+    /// Serializes the group of components from the world.
+    fn serialize_group<S: Serializer>(world: &World, serializer: S) -> Result<S::Ok, S::Error>;
+    /// Helper method for serializing the world.
```
Comment (WaDelma) [inline specs/src/group.rs]: Do we need to expose these helpers in the trait or are they only called in trait implementation? If they are not needed anywhere else then these helpers are just implementation detail and should probably removed.

**Comment by kvark** [comment]
Can we have the restructurization PR separate from the new functionality please?

**Comment by WaDelma** [comment]
You have to use workspace to include the custom derive as it needs to be it's own crate.

**Comment by Aceeri** [comment]
I mean, I could have another PR for it if we want, but it would either just be a workspace of "specs" and "specs_derive" (without functionality).

**Comment by torkleyy** [comment]
@Aceeri Maybe you could structure it the same way @ebkalderon did in their PR?

**Comment by Aceeri** [comment]
@torkleyy I feel it makes more sense to use a workspace given this is exactly the usecase for it (multiple crates in one repo).

**Comment by torkleyy** [comment]
@Aceeri They also implemented it as a workspace, see https://github.com/slide-rs/specs/pull/192/files

File: specs_derive/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
+# Specs
```
Comment (torkleyy) [inline specs_derive/README.md]: I don't think we need a second readme here (or at least, don't copy the specs README and just show a small example of component groups).

File: src/world.rs [hunk 5]
```diff
@@ -646,3 +660,67 @@ impl Default for World {
         }
     }
 }
+
+#[cfg(feature="serialize")]
+/// Structure used to serialize a world using a component group.
+pub struct WorldSerializer<'a, G> {
```
Comment (torkleyy) [inline src/world.rs]: Do we really need a struct for this?

Comment (Aceeri) [inline src/world.rs]: Yeah, if I implemented `Serialize` or `Deserialize` on the `World` itself, then it would make a new copy of the world.

Comment (torkleyy) [inline src/world.rs]: I was talking about just a function actually.

Comment (Aceeri) [inline src/world.rs]: Well, it ties in better with `serde`s traits and such this way rather than just have some special methods for it.

Comment (Aceeri) [inline src/world.rs]: I don't really mind either way since it would still be possible if it was a function instead of a struct.

**Comment by Aceeri** [comment]
Well uh, I have made a mistake so please ignore this for the next couple hours while I go get my backup.

**Comment by Aceeri** [comment]
@torkleyy For the `WorldSerializer`, I just realized why I originally had it. Some `serde` libraries don't expose `Serializer` in their interfaces which is annoying, but the only way to serialize with them is to pass in a type that implements `Serialize`.

File: src/world.rs [hunk 6]
```diff
@@ -780,3 +794,66 @@ mod tests {
...
+    entities: &'a [Entity],
+    phantom: PhantomData<G>,
+}
+
+#[cfg(feature="serialize")]
+impl<'a, G> WorldDeserializer<'a, G> {
+    /// Creates a new world deserializer out of a world and a list of entities.
+    ///
+    /// The list of entities will be used to merge into the component storages.
+    pub fn new(world: &'a mut World, entities: &'a [Entity]) -> WorldDeserializer<'a, G> {
```
Comment (torkleyy) [inline src/world.rs]: I wonder if it really makes sense to pass a slice of entities. Is there any case where you want to deserialize components and insert them for existing entities? I mean can you really ensure this works reliable? Which brings me to the question if it should be allowed for an entity to not have a component of the component group.

**Comment by torkleyy** [comment]
@Aceeri I see. It's up to you how you implement it then.

File: src/world.rs (hunk 6 above)
Comment (Aceeri) [inline src/world.rs]: > Is there any case where you want to deserialize components and insert them for existing entities? 

I was mainly thinking stuff like prefabs here, so a prefab could be deserialized into an existing entity.

> Which brings me to the question if it should be allowed for an entity to not have a component of the component group.

Well, if we disallowed that then every entity would have to have every component in that component group if we wanted to serialize the world properly.

Comment (torkleyy) [inline src/world.rs]: > I was mainly thinking stuff like prefabs here, so a prefab could be deserialized into an existing entity.

Exactly, so a new entity has to be created.

Comment (Aceeri) [inline src/world.rs]: What if you want to put in multiple prefabs for a single entity though?

**Comment by Aceeri** [comment]
Closing this for now since I'm not sure if it is worth the complexity yet.

**Pull request closed without merging**